tests/Integration/Behaviour/Features/Context/Domain/CartRuleFeatureContext.php

```diff
@@ -0,0 +1,640 @@
+                    $discountApplicationType
+                ));
+        }
+
+        if ($hasError) {
+            throw new RuntimeException(sprintf(
+                'Cart rule discount application type "%s" was expected',
+                $discountApplicationType
+            ));
+        }
+    }
+
+    /**
+     * @param string $nameInDefaultLanguage
+     * @param bool $highlightInCart
+     * @param bool $allowPartialUse
+     * @param int $priority
+     * @param bool $isActive
+     * @param DateTime $validFrom
+     * @param DateTime $validTo
+     * @param int $totalQuantity
+     * @param int $quantityPerUser
+     * @param CartRuleActionInterface $cartRuleAction
+     * @param float|null $minimumAmount
+     * @param int $minimumAmountCurrencyId
+     * @param bool $minimumAmountTaxIncluded
+     * @param bool $minimumAmountShippingIncluded
+     * @param string|null $code
+     * @param string|null $description
+     * @param int|null $customerId
+     * @param string|null $discountApplicationType
+     * @param int|null $discountProductId
+     *
+     * @return CartRuleId
+     *
+     * @throws CartRuleConstraintException
+     * @throws DomainConstraintException
+     */
+    private function createCartRule(
+        string $nameInDefaultLanguage,
+        bool $highlightInCart,
+        bool $allowPartialUse,
+        int $priority,
+        bool $isActive,
+        DateTime $validFrom,
+        DateTime $validTo,
+        int $totalQuantity,
+        int $quantityPerUser,
+        CartRuleActionInterface $cartRuleAction,
+        float $minimumAmount,
+        int $minimumAmountCurrencyId,
+        bool $minimumAmountTaxIncluded,
+        bool $minimumAmountShippingIncluded,
+        string $code = null,
+        string $description = null,
+        int $customerId = null,
+        string $discountApplicationType = null,
+        int $discountProductId = null
+    ) {
+        $defaultLanguageId = Configuration::get('PS_LANG_DEFAULT');
+
+        $command = new AddCartRuleCommand(
+            [$defaultLanguageId => $nameInDefaultLanguage],
+            $highlightInCart,
+            $allowPartialUse,
+            $priority,
+            $isActive,
+            $validFrom,
+            $validTo,
+            $totalQuantity,
+            $quantityPerUser,
+            $cartRuleAction,
+            $minimumAmount,
+            $minimumAmountCurrencyId,
+            !$minimumAmountTaxIncluded,
+            !$minimumAmountShippingIncluded
+        );
+
+        null !== $code ? $command->setCode($code) : null;
+        null !== $description ? $command->setDescription($description) : null;
+        null !== $customerId ? $command->setCustomerId($customerId) : null;
+        null !== $discountApplicationType ? $command->setDiscountApplicationType($discountApplicationType) : null;
+        null !== $discountProductId ? $command->setDiscountProductId($discountProductId) : null;
```

Comment:
This one is funny. :smile: Even though it seems shorter and cooler, it takes more brain power to understand compared to simple if. :smile:

Reply:
-2 lines per condition :older_man:

Reply:
should I refactor it? :smile:

Reply:
:man_shrugging:

Reply:
fixed